pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -396,3 +433,62 @@ func validatePipelineResults(results []PipelineResult) error {
 	}
 	return nil
 }
+
+func ValidateTasksAndFinallySection(ps *PipelineSpec) *apis.FieldError {
+	if len(ps.Finally) != 0 && len(ps.Tasks) == 0 {
+		return apis.ErrInvalidValue(fmt.Sprintf("spec.tasks is empty but spec.finally has %d tasks", len(ps.Finally)), "spec.finally")
+	}
+	return nil
+}
+
+func ValidateFinalTasks(finalTasks []PipelineTask) *apis.FieldError {
+	for _, f := range finalTasks {
+		if len(f.RunAfter) != 0 {
+			return apis.ErrInvalidValue(fmt.Sprintf("no runAfter allowed under spec.finally, final task %s has runAfter specified", f.Name), "spec.finally")
+		}
+		if len(f.Conditions) != 0 {
+			return apis.ErrInvalidValue(fmt.Sprintf("no conditions allowed under spec.finally, final task %s has conditions specified", f.Name), "spec.finally")
+		}
+	}
+
+	if err := validateTaskResultReference(finalTasks); err != nil {
+		return err
+	}
+
+	if err := validateTasksInputFrom(finalTasks); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateTaskResultReference(tasks []PipelineTask) *apis.FieldError {
+	for _, t := range tasks {
+		for _, p := range t.Params {
+			expressions, ok := GetVarSubstitutionExpressionsForParam(p)
+			if ok {
+				if LooksLikeContainsResultRefs(expressions) {
+					return apis.ErrInvalidValue(fmt.Sprintf("no task result allowed under params,"+
+						"final task param %s has set task result as its value", p.Name), "spec.finally.task.params")
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func validateTasksInputFrom(tasks []PipelineTask) *apis.FieldError {
```

Comment:
i think it would make sense to have unit tests for these as well - which based on our conversation with @vdemeester is an indication that this function probably belongs somewhere else - do you think you could make it work to have a function on PipelineTask that validates this? e.g. PipelineTask.Validate()? you could even have PipelineTask.Validate(), PipelineTask.ValidateFinally(), PipelineTask.ValidateNonFinally()

Reply:
and in that case i'd probably have way less test cases in TestValidatePipelineWithFinalTasks since we wouldnt have to cover absolutely everything, just enough to be confident the validation is being called

Reply:
breaking it up like that will make it easier to see what is covered and what isnt b/c you can look at it one function + set of unit tests at a time

Reply:
yup will look into it and create a separate PR, those changes would be independent of `finally`